File: src/api/diff.ts
import { invoke } from "@tauri-apps/api/core";
import type { ChangedFile, CommitChangedFile, FileDiff } from "@/types";

export const getChangedFiles = (worktreePath: string) => invoke<ChangedFile[]>("get_changed_files", { worktreePath });

export const getFileDiff = (worktreePath: string, filePath: string, staged: boolean) =>
  invoke<FileDiff>("get_file_diff", { worktreePath, filePath, staged });

export const getCommitChangedFiles = (worktreePath: string, sha: string) =>
  invoke<CommitChangedFile[]>("get_commit_changed_files", { worktreePath, sha });

export const getCommitFullDiff = (worktreePath: string, sha: string) =>
  invoke<FileDiff[]>("get_commit_full_diff", { worktreePath, sha });

export const getCommitFileDiff = (worktreePath: string, sha: string, filePath: string) =>
  invoke<FileDiff>("get_commit_file_diff", { worktreePath, sha, filePath });

export const getRangeChangedFiles = (worktreePath: string, fromSha: string, toSha: string) =>
  invoke<CommitChangedFile[]>("get_range_changed_files", { worktreePath, fromSha, toSha });

export const getRangeFullDiff = (worktreePath: string, fromSha: string, toSha: string) =>
  invoke<FileDiff[]>("get_range_full_diff", { worktreePath, fromSha, toSha });

export const getRangeFileDiff = (worktreePath: string, fromSha: string, toSha: string, filePath: string) =>
  invoke<FileDiff>("get_range_file_diff", { worktreePath, fromSha, toSha, filePath });

File: src-tauri/src/api/diff.rs
use std::path::PathBuf;

use crate::domain::diff::{CommitChangedFile, FileDiff};
use crate::domain::error::Result;
use crate::domain::worktree::ChangedFile;
use crate::service;

#[tauri::command]
pub async fn get_changed_files(worktree_path: String) -> Result<Vec<ChangedFile>> {
    super::blocking(move || service::diff::get_changed_files(&PathBuf::from(worktree_path))).await
}

#[tauri::command]
pub async fn get_file_diff(
    worktree_path: String,
    file_path: String,
    staged: bool,
) -> Result<FileDiff> {
    super::blocking(move || {
        service::diff::get_file_diff(&PathBuf::from(worktree_path), &file_path, staged)
    })
    .await
}

#[tauri::command]
pub async fn get_commit_changed_files(
    worktree_path: String,
    sha: String,
) -> Result<Vec<CommitChangedFile>> {
    super::blocking(move || {
        service::diff::get_commit_changed_files(&PathBuf::from(worktree_path), &sha)
    })
    .await
}

#[tauri::command]
pub async fn get_commit_full_diff(worktree_path: String, sha: String) -> Result<Vec<FileDiff>> {
    super::blocking(move || {
        service::diff::get_commit_full_diff(&PathBuf::from(worktree_path), &sha)
    })
    .await
}

#[tauri::command]
pub async fn get_commit_file_diff(
    worktree_path: String,
    sha: String,
    file_path: String,
) -> Result<FileDiff> {
    super::blocking(move || {
        service::diff::get_commit_file_diff(&PathBuf::from(worktree_path), &sha, &file_path)
    })
    .await
}

#[tauri::command]
pub async fn get_range_changed_files(
    worktree_path: String,
    from_sha: String,
    to_sha: String,
) -> Result<Vec<CommitChangedFile>> {
    super::blocking(move || {
        service::diff::get_range_changed_files(&PathBuf::from(worktree_path), &from_sha, &to_sha)
    })
    .await
}

#[tauri::command]
pub async fn get_range_full_diff(
    worktree_path: String,
    from_sha: String,
    to_sha: String,
) -> Result<Vec<FileDiff>> {
    super::blocking(move || {
        service::diff::get_range_full_diff(&PathBuf::from(worktree_path), &from_sha, &to_sha)
    })
    .await
}

#[tauri::command]
pub async fn get_range_file_diff(
    worktree_path: String,
    from_sha: String,
    to_sha: String,
    file_path: String,
) -> Result<FileDiff> {
    super::blocking(move || {
        service::diff::get_range_file_diff(
            &PathBuf::from(worktree_path),
            &from_sha,
            &to_sha,
            &file_path,
        )
    })
    .await
}
